core/module_specifier.rs

```diff
@@ -15,32 +76,31 @@ impl ModuleSpecifier {
   pub fn resolve(
     specifier: &str,
     base: &str,
-  ) -> Result<ModuleSpecifier, url::ParseError> {
+  ) -> Result<ModuleSpecifier, ResolveError> {
+    use ResolveError::*;
+
     // 1. Apply the URL parser to specifier. If the result is not failure, return
     //    the result.
-    // let specifier = parse_local_or_remote(specifier)?.to_string();
-    if let Ok(url) = Url::parse(specifier) {
-      return Ok(ModuleSpecifier(url));
+    if let Ok(module_specifier) = ModuleSpecifier::resolve_absolute(specifier) {
+      return Ok(module_specifier);
     }
 
-    // 2. If specifier does not start with the character U+002F SOLIDUS (/), the
-    //    two-character sequence U+002E FULL STOP, U+002F SOLIDUS (./), or the
-    //    three-character sequence U+002E FULL STOP, U+002E FULL STOP, U+002F
-    //    SOLIDUS (../), return failure.
-    if !specifier.starts_with('/')
-      && !specifier.starts_with("./")
-      && !specifier.starts_with("../")
+    // 2. If specifier does not start with the character U+002F SOLIDUS (/),
+    //    the two-character sequence U+002E FULL STOP, U+002F SOLIDUS (./),
+    //    or the three-character sequence U+002E FULL STOP, U+002E FULL STOP,
+    //    U+002F SOLIDUS (../), return failure.
+    if !(specifier.starts_with('/')
+      || specifier.starts_with("./")
+      || specifier.starts_with("../"))
     {
-      // TODO This is (probably) not the correct error to return here.
-      // TODO: This error is very not-user-friendly
-      return Err(url::ParseError::RelativeUrlWithCannotBeABaseBase);
+      Err(UnqualifiedRelativePath)?
     }
 
-    // 3. Return the result of applying the URL parser to specifier with base URL
-    //    as the base URL.
-    let base_url = Url::parse(base)?;
-    let u = base_url.join(&specifier)?;
-    Ok(ModuleSpecifier(u))
+    // 3. Return the result of applying the URL parser to specifier with base
+    //    URL as the base URL.
+    let base = Url::parse(base).map_err(InvalidBaseUrl)?;
```

Comment:
Not necessary for this PR, but let's move to pass in `base` as a Url too and not as a string.

Reply:
:+1: